# Práctica 2: Estimación e Inferencia en Modelos de Regresión Lineales

## Cuestiones

In [144]:
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
import matplotlib.pylab as plt
from wooldridge import *
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf
import matplotlib.pylab as plt
import statsmodels.stats.api as sms
from wooldridge import *
#dataWoo("gpa3", description=True)
import statsmodels.stats.outliers_influence as oi #Para el test RESET de Ramsey
import statsmodels.stats.diagnostic as diagn #Para el test de Harvey-Collier


dataWoo('hprice1',description=True)

name of dataset: hprice1
no of variables: 10
no of observations: 88

+----------+------------------------------+
| variable | label                        |
+----------+------------------------------+
| price    | house price, $1000s          |
| assess   | assessed value, $1000s       |
| bdrms    | number of bdrms              |
| lotsize  | size of lot in square feet   |
| sqrft    | size of house in square feet |
| colonial | =1 if home is colonial style |
| lprice   | log(price)                   |
| lassess  | log(assess                   |
| llotsize | log(lotsize)                 |
| lsqrft   | log(sqrft)                   |
+----------+------------------------------+

Collected from the real estate pages of the Boston Globe during 1990.
These are homes that sold in the Boston, MA area.


1. Utilizando la base de datos **hprice1**:
    - Estimar el modelo $\log {\rm price} = \beta_0 + \beta_1 \log {\rm lotsize} + \beta_2 \log{\rm sqrft} + \beta_3 {\rm bdrms} + u$. Interpretar los resultados obtenidos.
    - Obtener la predicción de $\log {\rm price}$ cuando ${\rm lotsize}=20000$, ${\rm sqrft}=2500$ y ${\rm bdrms}=4$.
    - Para explicar la variación de ${\rm price}$, se considera también el modelo: ${\rm price} = \beta^\prime_0 + \beta_1^\prime {\rm sqrtft} + \beta_2^\prime {\rm bdrms} + u^\prime$. ¿Que modelo es más útil?

In [145]:
datos=dataWoo("hprice1")

y=datos["lprice"]
X=datos[["llotsize", "lsqrft", "bdrms"]]

mco1 = sm.OLS(y, sm.add_constant(X)).fit()
mco1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 lprice   R-squared:                       0.643
Model:                            OLS   Adj. R-squared:                  0.630
Method:                 Least Squares   F-statistic:                     50.42
Date:                Sat, 18 Dec 2021   Prob (F-statistic):           9.74e-19
Time:                        13:53:58   Log-Likelihood:                 25.861
No. Observations:                  88   AIC:                            -43.72
Df Residuals:                      84   BIC:                            -33.81
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.2970      0.651     -1.992      0.050      -2.592      -0.002
llotsize       0.1680      0.038      4.388      0.000       0.092       0.244
lsqrft         0.7002      0.093      7.540      0.000       0.516       0.885
bdrms          0.0370      0.028      1.342      0.183      -0.018       0.092
==============================================================================
Omnibus:                       12.060   Durbin-Watson:                   2.089
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               34.889
Skew:                          -0.188   Prob(JB):                     2.65e-08
Kurtosis:                       6.062   Cond. No.                         410.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

--> La constante tiene un coeficiente de -1.29 lo que implica que cada incremento de este provoca un decremento de lprice en -1.29

--> La beta[1] llotsize tiene un coeficiente de 0.1680 lo que implica que cada incremento de este provoca un incremento de lprice en 0.1680

--> La beta[2] lsqrft tiene un coeficiente de 0.7002 lo que implica que cada incremento de este provoca un incremento de lprice en 0.7002

--> La beta[3] bdrms tiene un coeficiente de 0.0370 lo que implica que cada incremento de este provoca un incremento de lprice en 0.0370
    --> No es significativa

In [146]:
beta = mco1.params
prediccion = beta[0] + beta[1]*20000 + beta[2]*2500 + beta[3]*4
print("La prediccion es: ", prediccion)
#dime que pasa falin?
# Si es significativa afecta mas en los modelos es decir en la y

La prediccion es:  5108.765860093702


In [147]:
y1=datos["price"]
X1=datos[["sqrft", "bdrms"]]

mco1_aux = sm.OLS(y1, sm.add_constant(X1)).fit()
mco1_aux.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.632
Model:                            OLS   Adj. R-squared:                  0.623
Method:                 Least Squares   F-statistic:                     72.96
Date:                Sat, 18 Dec 2021   Prob (F-statistic):           3.57e-19
Time:                        13:53:58   Log-Likelihood:                -488.00
No. Observations:                  88   AIC:                             982.0
Df Residuals:                      85   BIC:                             989.4
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -19.3150     31.047     -0.622      0.536     -81.044      42.414
sqrft          0.1284      0.014      9.291      0.000       0.101       0.156
bdrms         15.1982      9.484      1.603      0.113      -3.658      34.054
==============================================================================
Omnibus:                       25.221   Durbin-Watson:                   1.858
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               44.973
Skew:                           1.122   Prob(JB):                     1.72e-10
Kurtosis:                       5.689   Cond. No.                     9.85e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.85e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

--> El primero es mas util, debido al condition numbre, r-squared... etc.

2. Usando los datos de **vote1**:
    - Estimar el modelo $voteA = \beta_0 + \beta_1 \log(expendA) + \beta_2 \log(expendB) + \beta_3 prtystrA + u$.
    - ¿Cuál es la interpretación de $\beta_1$?
    - ¿Los gastos de A afectan a los resultados de las elecciones? ¿Y los gastos de B?
    - Analizar la correcta especificación *lineal* del modelo con los Tests de Harvey-Collier y Reset de Ramsey.
    - Establecer la hipótesis nula de que un aumento de $1\%$ de aumento en los gastos de A es compensado por un aumento del $1\%$ en los gastos de B. Contrastar estadísticamente.

In [148]:
#dataWoo("vote1", description=True)

datos=dataWoo("vote1")
y=datos["voteA"]
X=datos[["lexpendA", "lexpendB", "prtystrA"]]

mco2 = sm.OLS(y, sm.add_constant(X)).fit()
mco2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.793
Model:                            OLS   Adj. R-squared:                  0.789
Method:                 Least Squares   F-statistic:                     215.2
Date:                Sat, 18 Dec 2021   Prob (F-statistic):           1.76e-57
Time:                        13:53:58   Log-Likelihood:                -596.86
No. Observations:                 173   AIC:                             1202.
Df Residuals:                     169   BIC:                             1214.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         45.0789      3.926     11.481      0.000      37.328      52.830
lexpendA       6.0833      0.382     15.919      0.000       5.329       6.838
lexpendB      -6.6154      0.379    -17.463      0.000      -7.363      -5.868
prtystrA       0.1520      0.062      2.450      0.015       0.030       0.274
==============================================================================
Omnibus:                        8.900   Durbin-Watson:                   1.604
Prob(Omnibus):                  0.012   Jarque-Bera (JB):                8.832
Skew:                           0.493   Prob(JB):                       0.0121
Kurtosis:                       3.505   Cond. No.                         344.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

*** Apartado B ***

--> El beta[1] "lexpendA" es el logaritmo de lo que se gasta el democrata A, su coeficiente es 6.0833, lo que implica que cada incremento de esta variable
implica un incremento de los votos en A en 6,0833 unidades.

--> Ademas, podemos ver que su p-valor es 0, por lo que es significativa a nivel individual y por tanto es importante para el modelo.

*** Apartado C ***

--> Si afectan por lo que he dicho en el apartado b

--> Y el "lexpendB" tambien afectaria a los votos pero de manera negativa ya que su coef es -6.6154 y su pvalor es significativo

In [149]:
import statsmodels.stats.diagnostic as diagn
import statsmodels.stats.api as sms
from scipy import stats

skip = len(mco2.params)  # bug in linear_harvey_collier
rr = sms.recursive_olsresiduals(mco2, skip=skip, alpha=0.95, order_by=None)

print("Test de Harvey Collier: ", stats.ttest_1samp(rr[3][skip:], 0), "--> No es lineal")

rr=oi.reset_ramsey(mco2, degree=2)
print("Test de RESET de Ramsey: ", rr, "--> Si es lineal")



Test de Harvey Collier:  Ttest_1sampResult(statistic=-5.966484195029227, pvalue=1.3957961709458505e-08) --> No es lineal
Test de RESET de Ramsey:  <F test: F=array([[0.24096435]]), p=0.624151071875373, df_denom=168, df_num=1> --> Si es lineal


In [150]:
# Como solo queremos incrementar el valor porcentual de la beta[1]-->"lexpendA" y la beta[2]-->"lexpendB" los ponemos a 1%

R=[[0,1,1,0]]

print(mco2.f_test(R).summary())
print("--> Como el valor es cercano a 0, se conpensan los valores, por lo tanto lo hemos comprobado estadisticamente")
print("--> Como el p-valor > 0.05 no rechazamos la hipotesis nula")

<F test: F=array([[0.99630884]]), p=0.3196323305021531, df_denom=169, df_num=1>
--> Como el valor es cercano a 0, se conpensan los valores, por lo tanto lo hemos comprobado estadisticamente
--> Como el p-valor > 0.05 no rechazamos la hipotesis nula








3. Con los datos disponibles en **mlb1** se pretende analizar el efecto de diversas características de los jugadores de la liga de béisbol americana en su salario. Se propone el siguiente modelo:
$$
\log (salary) = \beta_0 + \beta_1 years + \beta_2 gamesyr + \beta_3 bavg + \beta_4 hrunsyr + \beta_5 rbisyr + \beta_6 runsyr + \beta_7 fldperc + \beta_8 allstar + \beta_9 black + \beta_{10} hispan + \beta_{11} black \times percblck + \beta_{12} hispan \times perchisp + u
$$
    - Estimar el modelo.
    - ¿Son las variables $black$, $hispan$, $black\times percblck$ y $hispan\times perchisp$ conjuntamente significativas?
    - ¿Cómo interpretarías los coeficientes de las variables "raciales"?
    
    
    
    
    
    
    

In [151]:
datos=dataWoo("mlb1")
datos=datos.dropna(subset=['blckpb']) # Quita los elementos que tiene argumento nulos
y=datos['lsalary']
x=datos[[ 'years','gamesyr','bavg', 'hrunsyr','rbisyr' ,'runsyr', 'fldperc' , 'allstar' ,'black','hispan' , 'blckpb','hispph']]

mco3= sm.OLS(y, sm.add_constant(x) ).fit()

mco3.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                lsalary   R-squared:                       0.638
Model:                            OLS   Adj. R-squared:                  0.624
Method:                 Least Squares   F-statistic:                     46.48
Date:                Sat, 18 Dec 2021   Prob (F-statistic):           1.31e-62
Time:                        13:53:59   Log-Likelihood:                -350.12
No. Observations:                 330   AIC:                             726.2
Df Residuals:                     317   BIC:                             775.6
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         10.3437      2.183      4.739      0.000       6.050      14.638
years          0.0673      0.013      5.224      0.000       0.042       0.093
gamesyr        0.0089      0.003      2.624      0.009       0.002       0.016
bavg           0.0009      0.002      0.625      0.533      -0.002       0.004
hrunsyr        0.0146      0.016      0.889      0.375      -0.018       0.047
rbisyr         0.0045      0.008      0.593      0.553      -0.010       0.019
runsyr         0.0072      0.005      1.577      0.116      -0.002       0.016
fldperc        0.0011      0.002      0.513      0.609      -0.003       0.005
allstar        0.0075      0.003      2.621      0.009       0.002       0.013
black         -0.1980      0.125     -1.578      0.116      -0.445       0.049
hispan        -0.1900      0.153     -1.241      0.215      -0.491       0.111
blckpb         0.0125      0.005      2.509      0.013       0.003       0.022
hispph         0.0201      0.010      2.051      0.041       0.001       0.039
==============================================================================
Omnibus:                        1.828   Durbin-Watson:                   1.341
Prob(Omnibus):                  0.401   Jarque-Bera (JB):                1.709
Skew:                           0.012   Prob(JB):                        0.425
Kurtosis:                       3.352   Cond. No.                     5.65e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.65e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

--> Como la probabilidad de f-stadistic: 1.31e-62 < 0.05 podemos decir que nuestro modelo es fiable y que tiene una significacion a nivel global o conjunta.

--> Los coeficientes de las variables raciales son negativas, queriendonos decir que si somos negros o hispanos, nuestro salario decrementa.

4. Utilizar los datos de **gpa3** para analizar los promedios generales de calificaciones ($cumgpa$) de atletas universitarios en base a la puntuación del examen de acceso a la universidad ($sat$), el percentil alcanzado en bachiller ($hsperc$) y las horas de cursos universitarios ($totrs$):
$$
cumgpa = \beta_0 + \beta_1 sat + \beta_2 hsperc + \beta_3 tothrs + u
$$
    - Estimar el modelo anterior e interpreta los resultados.
    - ¿Está el modelo adecuadamente espeficicado de forma lineal según los tests de Harvey-Collier y RESET de Ramsey?
    - Para determinar si hay diferencias en las calificaciones entre hombres y mujeres del modelo anterior, considera el modelo con interacciones:
$$
cumgpa = \delta_0 + \delta_1 sat + \delta_2 hsperc + \delta_3 tothrs + \delta_4 female + \delta_5 female\times sat + \delta_6 female \times hsperc + \delta_7 female \times tothrs +  u
$$  
Estima el modelo e interpreta los coeficientes obtenidos.
    - Contrasta la hipótesis $H_0: \delta_4=\delta_5=\delta_6=\delta_7=0$. ¿Qué se puede deducir de lo obtenido?

In [152]:
datos=dataWoo("gpa3")

y=datos["cumgpa"]
X=datos[["sat", "hsperc", "tothrs"]]

mco4= sm.OLS(y, sm.add_constant(X) ).fit()

mco4.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 cumgpa   R-squared:                       0.235
Model:                            OLS   Adj. R-squared:                  0.232
Method:                 Least Squares   F-statistic:                     74.72
Date:                Sat, 18 Dec 2021   Prob (F-statistic):           3.87e-42
Time:                        13:53:59   Log-Likelihood:                -932.28
No. Observations:                 732   AIC:                             1873.
Df Residuals:                     728   BIC:                             1891.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9291      0.229      4.065      0.000       0.480       1.378
sat            0.0009      0.000      4.343      0.000       0.000       0.001
hsperc        -0.0064      0.002     -4.069      0.000      -0.009      -0.003
tothrs         0.0120      0.001     12.860      0.000       0.010       0.014
==============================================================================
Omnibus:                       12.916   Durbin-Watson:                   2.000
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               21.316
Skew:                           0.087   Prob(JB):                     2.35e-05
Kurtosis:                       3.818   Cond. No.                     6.53e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.53e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### Como las variables son significativas a nivel global, podemos aceptar la hipotesis de todos los sigmas = 0

In [153]:
import statsmodels.stats.diagnostic as diagn
import statsmodels.stats.api as sms
from scipy import stats

skip = len(mco4.params)  # bug in linear_harvey_collier
rr = sms.recursive_olsresiduals(mco2, skip=skip, alpha=0.95, order_by=None)

print("Test de Harvey Collier: ", stats.ttest_1samp(rr[3][skip:], 0), "--> No es lineal")

rr=oi.reset_ramsey(mco4, degree=2)
print("Test de RESET de Ramsey: ", rr, "--> No es lineal")

Test de Harvey Collier:  Ttest_1sampResult(statistic=-5.966484195029227, pvalue=1.3957961709458505e-08) --> No es lineal
Test de RESET de Ramsey:  <F test: F=array([[9.92329549]]), p=0.0016988604291086357, df_denom=727, df_num=1> --> No es lineal


In [154]:
datos["fsat"]=datos["female"]*datos['sat']
datos["fhperc"]=datos["female"]*datos['hsperc']
datos["ftot"]=datos["female"]*datos['tothrs']

y1=datos["cumgpa"]
X1=datos[["sat", "hsperc", "tothrs",'female','fsat','fhperc','ftot']]

mco4_aux= sm.OLS(y, sm.add_constant(X) ).fit()

mco4_aux.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 cumgpa   R-squared:                       0.235
Model:                            OLS   Adj. R-squared:                  0.232
Method:                 Least Squares   F-statistic:                     74.72
Date:                Sat, 18 Dec 2021   Prob (F-statistic):           3.87e-42
Time:                        13:53:59   Log-Likelihood:                -932.28
No. Observations:                 732   AIC:                             1873.
Df Residuals:                     728   BIC:                             1891.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9291      0.229      4.065      0.000       0.480       1.378
sat            0.0009      0.000      4.343      0.000       0.000       0.001
hsperc        -0.0064      0.002     -4.069      0.000      -0.009      -0.003
tothrs         0.0120      0.001     12.860      0.000       0.010       0.014
==============================================================================
Omnibus:                       12.916   Durbin-Watson:                   2.000
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               21.316
Skew:                           0.087   Prob(JB):                     2.35e-05
Kurtosis:                       3.818   Cond. No.                     6.53e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.53e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

5. Usando los datos de **sleep75**, y sabiendo que el modelo de interés es:
$$
sleep = \beta_0 + \beta_1 totwrk + \beta_2 educ + \beta_3 age + \beta_4 age^2 + \beta_5 yngkid + u
$$
    - Estimar el modelo por separado para hombres y mujeres y deducir su hay diferencias importantes entre los dos modelos estimados. Obtener las sumas de cuadrados de los residuos $SCR_H$ y $SCR_M$.
    - Estimar el modelo sin distinguir por género, y obtener $SCR$. Construir el estadístico:
$$
    F_{exp} = \frac{n-2k -2}{k+1} \frac{SCR - (SCR_H+SCR_M)}{SCR_H + SCR_M} \sim F_{k+1, n-2k-1}
$$
    donde $n$ es el número total de observaciones y $k$ el número de variables del modelo ($5$).
    - El estadístico $F_{exp}$ (**Test de Chow**) permite contrastar la hipótesis de que coeficientes de los modelos para hombres y mujeres coinciden. Comprobar si se rechaza o no esta hipótesis.

In [155]:
datos=dataWoo("sleep75")


datosH=datos[datos["male"]==1]

yH=datosH["sleep"]
XH=datosH[["totwrk", "educ", "age", "agesq", "yngkid"]]

mco5H= sm.OLS(yH, sm.add_constant(XH) ).fit()

e = mco5H.resid
mco5H.ssr

print("Suma de cuadrado de los residuos: ", sum(e**2))
mco5H.summary()

Suma de cuadrado de los residuos:  63763978.989306204


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sleep   R-squared:                       0.156
Model:                            OLS   Adj. R-squared:                  0.146
Method:                 Least Squares   F-statistic:                     14.59
Date:                Sat, 18 Dec 2021   Prob (F-statistic):           3.95e-13
Time:                        13:53:59   Log-Likelihood:                -2963.4
No. Observations:                 400   AIC:                             5939.
Df Residuals:                     394   BIC:                             5963.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3648.2083    310.039     11.767      0.000    3038.670    4257.747
totwrk        -0.1821      0.024     -7.438      0.000      -0.230      -0.134
educ         -13.0524      7.414     -1.760      0.079     -27.629       1.524
age            7.1566     14.320      0.500      0.618     -20.997      35.310
agesq         -0.0448      0.168     -0.266      0.791      -0.376       0.286
yngkid        60.3802     59.023      1.023      0.307     -55.659     176.419
==============================================================================
Omnibus:                       21.716   Durbin-Watson:                   2.061
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               42.802
Skew:                          -0.305   Prob(JB):                     5.08e-10
Kurtosis:                       4.482   Cond. No.                     4.71e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.71e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [156]:
datosM=datos[datos["male"]==0]

yM=datosM["sleep"]
XM=datosM[["totwrk", "educ", "age", "agesq", "yngkid"]]

mco5M= sm.OLS(yM, sm.add_constant(XM) ).fit()

e = mco5M.resid
mco5M.ssr

print("Suma de cuadrado de los residuos: ", sum(e**2))

mco5M.summary()

Suma de cuadrado de los residuos:  57288575.94098887


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sleep   R-squared:                       0.098
Model:                            OLS   Adj. R-squared:                  0.083
Method:                 Least Squares   F-statistic:                     6.495
Date:                Sat, 18 Dec 2021   Prob (F-statistic):           9.42e-06
Time:                        13:53:59   Log-Likelihood:                -2291.6
No. Observations:                 306   AIC:                             4595.
Df Residuals:                     300   BIC:                             4618.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       4238.7293    384.892     11.013      0.000    3481.299    4996.160
totwrk        -0.1399      0.028     -5.060      0.000      -0.194      -0.086
educ         -10.2051      9.589     -1.064      0.288     -29.075       8.665
age          -30.3566     18.531     -1.638      0.102     -66.824       6.110
agesq          0.3679      0.223      1.647      0.101      -0.072       0.807
yngkid      -118.2826     93.188     -1.269      0.205    -301.667      65.102
==============================================================================
Omnibus:                       54.909   Durbin-Watson:                   1.931
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              192.655
Skew:                          -0.730   Prob(JB):                     1.46e-42
Kurtosis:                       6.602   Cond. No.                     3.93e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.93e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [157]:
from scipy import stats

y=datos["sleep"]
X=datos[["totwrk", "educ", "age", "agesq", "yngkid"]]

mco5= sm.OLS(y, sm.add_constant(X) ).fit()

scrH=mco5H.ssr
scrM=mco5M.ssr

scr = mco5.ssr
n = mco5.nobs

k=5

Fexp= ((n-2*k-2)/(k-1)) * ((scr-(scrH+scrM))/(scrH+scrM))

alpha=0.05
Fteo= stats.f.ppf(1-alpha,k+1,n-2*k-1)

print("La F-teorico: ", Fteo)
print("La F-experimental: ", Fexp)

print("Rechazamos la hipotesis nula, ya que nuestro f-teorico es mas pequeño que nuestra f-experimental, por lo que los coeficiente de mujeres y hombres no coinciden")

La F-teorico:  2.1116070794184885
La F-experimental:  21386902583.00718
Rechazamos la hipotesis nula, ya que nuestro f-teorico es mas pequeño que nuestra f-experimental, por lo que los coeficiente de mujeres y hombres no coinciden


6. Aplicar el test de Chow del ejercicio anterior para determinar si existen diferencias entre géneros al analizar el modelo $\log (wage) = \beta_0 + \beta_1 educ + \beta_2 exper + \beta_3 exper^2 + \beta_4 tenure + \beta_5 tenure^2 + u$ con los datos de **wage1**.

In [158]:
datos=dataWoo("wage1")

y=datos["lwage"]
X=datos[["educ", "exper", "expersq", "tenure", "tenursq"]]

mco6= sm.OLS(y, sm.add_constant(X) ).fit()


datosM=datos[datos["female"]==1]

yM=datosM["lwage"]
XM=datosM[["educ", "exper", "expersq", "tenure", "tenursq"]]

mco6M= sm.OLS(yM, sm.add_constant(XM) ).fit()

datosH=datos[datos["female"]==0]

yH=datosH["lwage"]
XH=datosH[["educ", "exper", "expersq", "tenure", "tenursq"]]

mco6H= sm.OLS(yH, sm.add_constant(XH) ).fit()

scrH=mco6H.ssr
scrM=mco6M.ssr
scr = mco6.ssr
n = mco6.nobs

k=6

alpha=0.05
Fteo= stats.f.ppf(1-alpha,k+1,n-2*k-1)

#Test de chow
Fexp= ((n-2*k-2)/(k-1)) * ((scr-(scrH+scrM))/(scrH+scrM))

print("La F-teorico: ", Fteo)
print("La F-experimental: ", Fexp)

print("Rechazamos la hipotesis nula, ya que nuestro f-teorico es mas pequeño que nuestra f-experimental, por lo que los coeficiente de mujeres y hombres no coinciden")

La F-teorico:  2.1161718633941704
La F-experimental:  11939.100341009516
Rechazamos la hipotesis nula, ya que nuestro f-teorico es mas pequeño que nuestra f-experimental, por lo que los coeficiente de mujeres y hombres no coinciden


7. Utilizando **apple**:
    - Definir la variable binaria $eco=1$ si $ecolbs>0$ y $eco=0$ y si $ecolbs=0$. ¿Qué indica esa variable? ¿Cuál es la proporcion de familias que asegura que comprará manzanas ecoetiquetadas?
    - Estimar el modelo $eco = \beta_0 + \beta_1 ecoprc + \beta_2 regprc +\beta_3 faminc + \beta_4 hhsize + \beta_5 educ + \beta_6 age + u$. Interpretar $\beta_1$ y $\beta_2$.
    - ¿Son conjuntamente significativas las variables que no corresponden a precios del modelo? ¿Qué variable independiente, además de las relacionadas con precios, parecen tener unefecto más importante sobre la decisión de comprar manzanas ecológicas? ¿Tiene sentido?
    - Sustituir en el modelo la variable $faminc$ por $\log (faminc)$ y estimar el nuevo modelo. ¿Qué modelo se ajusta mejor?
    - Utilizar el Test de Chow para determinar si hay diferencias entre hombres y mujeres en su intención de comprar manzanas ecológicas.

In [159]:
print(dataWoo("apple", description=True))
datos=dataWoo("apple")

# La variable ecolbs el numero de manzanas etiquetadas
# La variable eco, si la manzana esta etiquetada o no
datos.loc[datos["ecolbs"] == 0, "eco"] = 0
datos.loc[datos['ecolbs'] > 0.01, 'eco'] = 1

print("La proporcion de familias que asegura que comprara manzanas ecoetiquetadas: ", np.mean(datos["eco"])*100, "%")

Y = datos["eco"]
X = datos[["ecoprc", "regprc", "faminc", "hhsize", "educ", "age"]]

mco7 = sm.OLS(Y, sm.add_constant(X) ).fit()
mco7.summary()

name of dataset: apple
no of variables: 17
no of observations: 660

+----------+---------------------------------+
| variable | label                           |
+----------+---------------------------------+
| id       | respondent identifier           |
| educ     | years schooling                 |
| date     | date:  month/day/year           |
| state    | home state                      |
| regprc   | price of regular apples         |
| ecoprc   | price of ecolabeled apples      |
| inseason | =1 if interviewed in Nov.       |
| hhsize   | household size                  |
| male     | =1 if male                      |
| faminc   | family income, thousands        |
| age      | in years                        |
| reglbs   | quantity regular apples, pounds |
| ecolbs   | quantity ecolabeled apples, lbs |
| numlt5   | # in household younger than 5   |
| num5_17  | # in household 5 to 17          |
| num18_64 | # in household 18 to 64         |
| numgt64  | # in household older than 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    eco   R-squared:                       0.110
Model:                            OLS   Adj. R-squared:                  0.102
Method:                 Least Squares   F-statistic:                     13.43
Date:                Sat, 18 Dec 2021   Prob (F-statistic):           2.18e-14
Time:                        13:53:59   Log-Likelihood:                -419.60
No. Observations:                 660   AIC:                             853.2
Df Residuals:                     653   BIC:                             884.6
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.4237      0.165      2.568      0.010       0.100       0.748
ecoprc        -0.8026      0.109     -7.336      0.000      -1.017      -0.588
regprc         0.7193      0.132      5.464      0.000       0.461       0.978
faminc         0.0006      0.001      1.042      0.298      -0.000       0.002
hhsize         0.0238      0.013      1.902      0.058      -0.001       0.048
educ           0.0248      0.008      2.960      0.003       0.008       0.041
age           -0.0005      0.001     -0.401      0.689      -0.003       0.002
==============================================================================
Omnibus:                     4015.360   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               69.344
Skew:                          -0.411   Prob(JB):                     8.75e-16
Kurtosis:                       1.641   Cond. No.                         724.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [160]:
datos["lfaminc"]=np.log(datos["faminc"])

Y = datos["eco"]
X = datos[["ecoprc", "regprc", "lfaminc", "hhsize", "educ", "age"]]

mco7_aux = sm.OLS(Y, sm.add_constant(X) ).fit()
mco7_aux.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    eco   R-squared:                       0.112
Model:                            OLS   Adj. R-squared:                  0.103
Method:                 Least Squares   F-statistic:                     13.67
Date:                Sat, 18 Dec 2021   Prob (F-statistic):           1.16e-14
Time:                        13:54:00   Log-Likelihood:                -418.94
No. Observations:                 660   AIC:                             851.9
Df Residuals:                     653   BIC:                             883.3
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.3038      0.179      1.697      0.090      -0.048       0.655
ecoprc        -0.8007      0.109     -7.326      0.000      -1.015      -0.586
regprc         0.7214      0.132      5.485      0.000       0.463       0.980
lfaminc        0.0445      0.029      1.550      0.122      -0.012       0.101
hhsize         0.0227      0.013      1.810      0.071      -0.002       0.047
educ           0.0231      0.008      2.733      0.006       0.006       0.040
age           -0.0004      0.001     -0.309      0.758      -0.003       0.002
==============================================================================
Omnibus:                     3113.689   Durbin-Watson:                   2.088
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               68.818
Skew:                          -0.412   Prob(JB):                     1.14e-15
Kurtosis:                       1.649   Cond. No.                         499.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

El modelo auxiliar con lfamine --> es mejor porque su R-squared es mayor

In [161]:
Y = datos["eco"]
X = datos[["ecoprc", "regprc", "faminc", "hhsize", "educ", "age"]]

mco7_aux = sm.OLS(Y, sm.add_constant(X) ).fit()

datosM=datos[datos["male"]==0]

yM=datosM["eco"]
XM=datosM[["ecoprc", "regprc", "faminc", "hhsize", "educ", "age"]]

mco7M_aux= sm.OLS(yM, sm.add_constant(XM) ).fit()

datosH=datos[datos["male"]==1]

yH=datosH["eco"]
XH=datosH[["ecoprc", "regprc", "faminc", "hhsize", "educ", "age"]]

mco7H_aux= sm.OLS(yH, sm.add_constant(XH) ).fit()

scrH=mco7H_aux.ssr
scrM=mco7M_aux.ssr
scr = mco7_aux.ssr
n = mco7_aux.nobs

k=6

alpha=0.05
Fteo= stats.f.ppf(1-alpha,k+1,n-2*k-1)

#Test de chow
Fexp= ((n-2*k-2)/(k-1)) * ((scr-(scrH+scrM))/(scrH+scrM))

print("La F-teorico: ", Fteo)
print("La F-experimental: ", Fexp)

print("Rechazamos la hipotesis nula, ya que nuestro f-teorico es mas pequeño que nuestra f-experimental, por lo que los coeficiente de mujeres y hombres no coinciden")

La F-teorico:  2.112532024286166
La F-experimental:  22163.243121763644
Rechazamos la hipotesis nula, ya que nuestro f-teorico es mas pequeño que nuestra f-experimental, por lo que los coeficiente de mujeres y hombres no coinciden
